packages/vite/src/node/server/middlewares/static.ts

```diff
@@ -107,40 +104,48 @@ export function serveRawFsMiddleware(
   }
 }
 
-export function isFileAccessAllowed(
+export function isFileServingAllowed(
   url: string,
-  { root, strict }: Required<FileSystemServeOptions>
+  server: ViteDevServer
 ): boolean {
-  return !strict || normalizePath(url).startsWith(root + path.posix.sep)
+  if (!server.config.server.fsServe.strict) return true
+
+  const file = cleanUrl(url)
+
+  if (server.moduleGraph.safeModulesPath.has(file)) {
+    return true
+  }
+
+  const normalizedUrl = normalizePath(file)
+  if (
+    server.config.server.fsServe.dirs.some((i) =>
+      normalizedUrl.startsWith(i + path.posix.sep)
+    )
+  )
+    return true
+
+  return false
```

Comment:
This can be squashed into one line

```suggestion
  return server.config.server.fsServe.dirs.some((i) =>
    normalizedUrl.startsWith(i + path.posix.sep)
  )
```